src/box/lua/connpool.lua

```diff
@@ -383,6 +467,60 @@ local function get_connection(opts)
     return nil, "connection to candidates failed"
 end
 
+-- This method looks for an active connection in the whole
+-- connection pool taking the specified priorities in account.
+--
+-- If there is no such connection the method tries to
+-- establish it with one of the matching replicas.
+--
+-- Used as a subroutine for the `call()` method.
+local function select_connection(opts)
+    assert(opts.mode == nil or opts.mode == 'ro' or opts.mode == 'rw' or
+           opts.mode == 'prefer_ro' or opts.mode == 'prefer_rw')
+
+    -- It's better to use the local instance to perform the call
+    -- faster. So prefer the local instance by default.
+    local prefer_local = true
+    if opts.prefer_local ~= nil then
+        prefer_local = opts.prefer_local
+    end
+
+    local filter_static_opts = {
+        groups = opts.groups,
+        replicasets = opts.replicasets,
+        instances = opts.instances,
+        labels = opts.labels,
+        roles = opts.roles,
+        sharding_roles = opts.sharding_roles,
+
+        -- The connection check isn't needed since it connects
+        -- to all of the candidates while we're trying to acquire
+        -- any candidate.
+        skip_connection_check = true,
+    }
+
+    -- At first we want to acquire the instances matching the
+    -- static (configuration) requirements.
+    local static_candidates = filter(filter_static_opts)
+
+    -- In case we don't need to prioritize the matching instances
+    -- we can wait for any connection available.
+    local need_priorities =
+        opts.mode == 'prefer_rw' or opts.mode == 'prefer_ro'
```

Comment:
What was not obvious for me: why we don't perform the dynamic filtering if the instances are to be prioritized. Then I looked over all the dynamic filters and understood that they don't shrink the list of candidates in these `prefer_local`, `prefer_ro`, `prefer_rw` cases.

But I would highlight it explicitly, because it looks like we forgot to perform the dynamic filtering.